### DISCUSSION TITLE
Attempting to output change set and their associated work items

### DISCUSSION THREAD

**Discussion by mrspaceman8** [rfennell/AzurePipelines#1521] [Q&A] [answered]
Hello, 
I am trying to output in a wiki page change sets and their associated work items in the form of:

build
-changeset 1
     work item 1
-changeset 2
     work item 2
     work item 3

I am using both the Release Notes Generator app and the Wiki File Updater. 
This is the code I am using in the template to output into the Wiki file:

**Global list of ChangeSets** ({{commits.length}})
{{#forEach commits}}
  {{#if isFirst}}
    **Associated commits:**
  {{/if}}
  ** [ChangeSetID {{remove this.id "C"}}](.../_versionControl/changeset/{{remove this.id "C"}})**
  Message: {{this.message}}
  Commited by: {{this.author.displayName}}
  FileCount: {{{this.changes.length}}}
  {{/forEach}}

{{#forEach this.workItems}}
   RelatedWorkItemID {{this.id}}
{{/forEach}}

I have tried this in two separate builds. One gives me an output in the following format:
build
-changeset 1
-changeset 2
-work item 1 (related to CS1)
-work item 2 (related to CS2)

Whilst the second one doesn't print the work items at all. 

I am relatively new to this, so I would appreciate any help you can give me :) 

Thank you

**Comment by rfennell** [comment]
I would suggest you watch the [video of the conference session](https://www.youtube.com/watch?v=xaV3dFoQdV8&t=332s) I did on this tool, as I think this might help

The other thing to do is look at the detailed log (set a variable `system.debug=true`  as this shows all the API calls being made and usually tells you what the issue is

**Reply by mrspaceman8** [reply]
Hi Richard, thank you for answering. i went through your video and I still couldn't find the answer to my question. I think I am missing some basic understanding of how you can link WI and CS together. I've tried this format but i am getting an empty result for the work item based on CS

{{#forEach commits}}
....
{{#if workitems}}
Work item: {{this.workitems.id}}
{{/if}}
{{/forEach}}

Is there even a way to do this?

**Reply by rfennell** [reply, marked as answer]
WI are associated to CS by either a comment in the commit message in the form #1234, or by adding the association in the IDE UX when you do the commit. So when the API looks for the differences between two successful runs of a build definition only WI associated CS between those runs are shown.

There is also some bespoke logic in the task to find parent WI, but this only works once the direct associated WI are found.

The way everything is build up is shown in the detailed debug log
